### ISSUE TITLE
Support binary download of spool content in ZOWE Explorer extender APIs 

### ISSUE THREAD

**Issue by VitGottwald** [zowe/vscode-extension-for-zowe#2060]
**Is your feature request related to a problem? Please describe.**

We have a process that stores binary data in spool files and we would like to use ZOWE Explorer API to download it in a VSCode extension.

**Describe the solution you'd like**

Support binary download of spool content in ZOWE Explorer extender APIs (at least through z/OSMF).

**Describe alternatives you've considered**

We have a working mechanism to download the content through FTP (without ZOWE Explorer), but we would like to support the scenario with ZOSMF as well.

**Additional context**

The parameters of https://github.com/zowe/vscode-extension-for-zowe/blob/main/packages/zowe-explorer-api/src/profiles/ZoweExplorerZosmfApi.ts#L325 do not provide the ability to select binary/text download. It also looks like it is not implemented in the ZOWE SDK for jobs https://github.com/zowe/zowe-cli/blob/master/packages/zosjobs/src/DownloadJobs.ts#L52 . However, ZOSMF itself seems to support this. From the [z/OSMF documentation](https://www.ibm.com/docs/en/zos/2.4.0?topic=interface-retrieve-contents-job-spool-file) the endpoint `/zosmf/restjobs/jobs/{correlator}/files/{nnn}/records` accepts  a query parameter `mode` with possible values `text/binary/record`.
...

**Comment by zFernand0** [comment]
Hey @VitGottwald,
The implementation of the binary (and record) optional flags in the CLI (and Node.js SDK) was published in version `v7.10.0`. 
Said enhancement was included in our Zowe Explorer API `v2.7.0`.
Here is a quick example on how to make use of it from a pure Zowe Explorer API perspective.
```javascript
const zowe_explorer_api = require("@zowe/zowe-explorer-api");
const imperative = require("@zowe/imperative");
(async () => {
  const profile_info = new imperative.ProfileInfo("zowe");
  await profile_info.readProfilesFromDisk();
  const default_profile = profile_info.getDefaultProfile("zosmf");
  const loaded_profile = imperative.ProfileInfo.profAttrsToProfLoaded(default_profile);
  const merged_arguments = profile_info.mergeArgsForProfile(default_profile, { getSecureVals: true });
  const session = imperative.ProfileInfo.createSession(merged_arguments.knownArgs);
  const zosmf_jes_api = new zowe_explorer_api.ZosmfJesApi(loaded_profile);
  zosmf_jes_api.session = session;

  await zosmf_jes_api.downloadSpoolContent({
    jobid: "JOB01234",
    jobname: "JOBNAME",
    binary: true,
    outDir: __dirname
  });
})();
```

---

That being said, we are working on exposing this to end users somehow.
Here are our thoughts:
1. A new context-menu option.
  ![image](https://user-images.githubusercontent.com/37381190/229905246-8ae368b7-2cab-4f53-9fcb-ab3e0144aadc.png)
  
2. Or automatically detecting that the downloaded spool content has non-printable characters. 
  ![image](https://user-images.githubusercontent.com/37381190/229905820-8c6712d1-d03a-4a4a-8d1e-3e49170c215a.png)

    This approach is something similar to VSCode complains depending the the file tags on USS files
    ![image](https://user-images.githubusercontent.com/37381190/229908862-de8d6944-805f-4e6a-b762-e0394858615e.png)

---
Here is a test VSIX file for approach number 1 😋 
[zowe-explorer-2060.zip](https://github.com/zowe/vscode-extension-for-zowe/files/11152684/zowe-explorer-2060.zip)

Please let us know if that's sufficient (for now), or if you would like us to investigate a bit more what it would take to do the auto-detection.

**Comment by VitGottwald** [comment]
This is a great news, thank you for following up and integrating the new CLI capability to ZE @zFernand0!

Lets go through a typical scenario and for the sake of simplicity lets assume that my job only has one step. With multiple steps, the same logic would apply.

1. I submit my job
2. the step inside the job produces two outputs
    * the first output file (SYSPRINT) is a standard text report about the operation of the utility
        * summary of the requested operation
        * progress of the processing
        * encountered errors
        * final success or failure messages
    * the second output file (SYSDATA) contains binary data that is the actual result of the operation
3. I find the job in the ZE job view. I expand it and I see 5 outputs
    * (1) JESMSGLG (text)
    * (2) JESJCL (text)
    * (3) JESYSMSG (text)
    * (4) SYSPRINT (text)
    * (5) SYSDATA (binary)
4. I check JESMSGLG (text) for any allocation errors, abends, or other operating system messages. When everything is fine there ...
5. I check the SYSPRINT (text) to see the report about the requested operation. Everything looks good here as well.
6. Now I would like to inspect the _binary output file SYSDATA_

I believe that the two options you offer:

* downloading as text, detecting non-printable characters and offering a re-download in binary and
* right-click and select download in binary from context menu

are very nice will address the need from UI perspective.

The crucial piece is that these options are available on **individual spool output files** rather than on the whole job. What do you think?

**Comment by zFernand0** [comment]
I think it makes perfect sense. :yum:

I felt a bit strange having to download the whole thing as binary because most DDs would just be garbled output except for the one (SYSDATA) that was actually meant to be binary.

I'll look into it further today.
Thanks for the concrete example!

**Comment by zFernand0** [comment]
Thanks for sharing a little more info regarding the use case
After looking into this a little more, there seems to be no way to stream spool content without having to construct the zosmf endpoint yourself 😢 

I'm looking into it though 😋

**Comment by zFernand0** [comment]
Hey @VitGottwald,
Hopefully this is the final update 😋 

After a quick update dependency update to the Zowe Explorer API, I'm able to stream the data in binary format 😋 

Here is a quick script that might help

```typescript
/** ========== Setup ============= */
const zowe_explorer_api = require("@zowe/zowe-explorer-api");
const imperative = require("@zowe/imperative");
const zowe_jobs_sdk = require("@zowe/zos-jobs-for-zowe-sdk");
const node_stream = require("stream");
(async () => {
  const profile_info = new imperative.ProfileInfo("zowe");
  await profile_info.readProfilesFromDisk();
  const default_profile = profile_info.getDefaultProfile("zosmf");
  const loaded_profile = imperative.ProfileInfo.profAttrsToProfLoaded(default_profile);
  const merged_arguments = profile_info.mergeArgsForProfile(default_profile, { getSecureVals: true });
  const session = imperative.ProfileInfo.createSession(merged_arguments.knownArgs);
  const zosmf_jes_api = new zowe_explorer_api.ZosmfJesApi(loaded_profile);
  zosmf_jes_api.session = session;
/** ========== End of Setup ============= */

  const jobid = "JOB01234";
  const jobname = "JOBNAME";
  const spools = (await zowe_jobs_sdk.GetJobs.getSpoolFiles(session, jobname, jobid));
  const spool = spools[spools.length - 1]; // selecting last spool for testing purposes

  // Creating a node_stream Writable
  const myStream = new node_stream.Writable({write: (chunk) => { console.log(chunk.toString()); }});

  // Creating a writeStream yourself (wither with Imperative of with node_fs)
  // const myStream = imperative.IO.createWriteStream("path/to/binary.jar");
  // const myStream = require("fs").createWriteStream("path/to/binary.jar");

  // Creating a Duplex (for socket-like read/write of data)
  // const myStream = new node_stream.Duplex({write: function (chunk){ this.push(chunk)}, read: () => {}});
  // myStream.on("data", (chunk) => console.log(chunk.toString()))

  await zosmf_jes_api.downloadSingleSpool({
  // await zowe_jobs_sdk.DownloadJobs.downloadSpoolContentCommon(session, {
    jobFile: spool,
    binary: true,
    stream: myStream
  });
})();
```

**Comment by VitGottwald** [comment]
This looks very promising, thank you @zFernand0 ! We will try it out and let you know. Is there a build we could take to do this?

**Comment by VitGottwald** [comment]
One more question out of curiosity at the moment. Since you managed to get the streams API work, did you also think what it would take to expose the endpoints through [web streams API](https://css-tricks.com/web-streams-everywhere-and-fetch-for-node-js/)?

**Comment by zFernand0** [comment]
Here is a tgz that you can experiment with.
[zowe-explorer-api-2.8.0-SNAPSHOT.tgz](https://github.com/zowe/vscode-extension-for-zowe/files/11314790/zowe-zowe-explorer-api-2.8.0-SNAPSHOT.tgz)
(here is a `.zip` in case you can't download the `.tgz` due to company policy 😋)
[zowe-explorer-api-2.8.0-SNAPSHOT.tgz.zip](https://github.com/zowe/vscode-extension-for-zowe/files/11314799/zowe-zowe-explorer-api-2.8.0-SNAPSHOT.tgz.zip)
These changes will be available on with Zowe Explorer 2.8.0 😋 


---

As far as adding support for web streams, it will require a bit more time in order to do it right.
This goes back to when the Imperative framework was built (around the time Node.js 8 was still in support) 😋 
We may require some fundamental changes to Imperative, although said changes are not impossible to make, just hard to get prioritized 😋 

Personally, I agree with most comments on that article regarding node streams being cumbersome 😅
